Provider crash when creating a `octopusdeploy_project_versioning_strategy` resource

**Describe the bug**
Provider crashes when trying to add a 'octopusdeploy_project_versioning_strategy'. Also unable to import it.

**Steps to reproduce**
Create a project, then add a 'octopusdeploy_project_versioning_strategy' resource.

**Expected behavior**
Being able to specify via terraform the versioning strategy. I'd be ok with importing the current value but it doesn't look like it's supported.

**Logs and other supporting information**

```
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-octopusdeploy_v0.39.0.exe plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0x10c33c5]

goroutine 47 [running]:
github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework.mapProjectVersioningStrategyToState(...)
        github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/resource_project_versioning_strategy.go:188
github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework.(*projectVersioningStrategyResource).Create(0xc00052e040, {0x17782f8, 0xc0008907e0}, {{{{0x1780058, 0xc0008915c0}, {0x1431260, 0xc000891320}}, {0x1784490, 0xc0005370e0}}, {{{0x1780058, ...}, ...}, ...}, ...}, ...)
        github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/resource_project_versioning_strategy.go:80 +0x705
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0xc00073c000, {0x17782f8, 0xc0008907e0}, 0xc00094f540, 0xc00094f518)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/fwserver/server_createresource.go:101 +0x578
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0xc00073c000, {0x17782f8, 0xc0008907e0}, 0xc0005fe370, 0xc00094f630)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/fwserver/server_applyresourcechange.go:57 +0x4aa
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0xc00073c000, {0x17782f8?, 0xc0008906f0?}, 0xc0005fe2d0)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/proto6server/server_applyresourcechange.go:55 +0x38e
github.com/hashicorp/terraform-plugin-mux/tf6muxserver.(*muxServer).ApplyResourceChange(0xc0002e4230, {0x17782f8?, 0xc000890420?}, 0xc0005fe2d0)
        github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf6muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc00027eaa0, {0x17782f8?, 0xc0003025d0?}, 0xc000124690)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/tf6server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x155a8a0, 0xc00027eaa0}, {0x17782f8, 0xc0003025d0}, 0xc000928380, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00054c800, {0x17782f8, 0xc000302540}, {0x1782480, 0xc00027ca80}, 0xc000148a20, 0xc000699ef0, 0x1f818b8, 0x0)
        google.golang.org/grpc@v1.64.1/server.go:1379 +0xdf8
google.golang.org/grpc.(*Server).handleStream(0xc00054c800, {0x1782480, 0xc00027ca80}, 0xc000148a20)
        google.golang.org/grpc@v1.64.1/server.go:1790 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.64.1/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 14
        google.golang.org/grpc@v1.64.1/server.go:1040 +0x125

Error: The terraform-provider-octopusdeploy_v0.39.0.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
````


**Environment and versions:**
- OS: Windows
- Octopus Server Version: Cloud
- Terraform Version: 

```
Terraform v1.8.3
on windows_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.39.0
+ provider registry.terraform.io/hashicorp/azurerm v3.103.1
+ provider registry.terraform.io/hashicorp/random v3.6.1
+ provider registry.terraform.io/octopusdeploylabs/octopusdeploy v0.39.0

Your version of Terraform is out of date! The latest version
is 1.10.5. You can update by downloading from https://www.terraform.io/downloads.html
```

--- a/.github/workflows/integration-tests.yaml
+++ b/.github/workflows/integration-tests.yaml
@@ -99,7 +99,7 @@ jobs:
           OCTOTESTRETRYCOUNT: 1
       - name: Upload test artifacts
         if: always()
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: junit-test-summary-${{ matrix.index }}
           path: node-summary.xml
@@ -117,7 +117,7 @@ jobs:
       - uses: actions/setup-node@v3
 
       - name: Download artifacts
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@v4
 
       - name: Install junit-report-merger
         run: npm install -g junit-report-merger
@@ -142,7 +142,7 @@ jobs:
           "junit-test-summary-14/*.xml"
 
       - name: Upload test artifacts
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: junit-test-summary
           path: ./junit-test-summary.xml

--- a/docs/resources/project_versioning_strategy.md
+++ b/docs/resources/project_versioning_strategy.md
@@ -60,7 +60,21 @@ resource "octopusdeploy_deployment_process" "process" {
   depends_on  = [octopusdeploy_project.tp]
 }
 
-resource "octopusdeploy_project_versioning_strategy" "tp" {
+######
+# NOTE: You can use either template or donor_package, not both
+######
+resource "octopusdeploy_project_versioning_strategy" "using_template" {
+  project_id = octopusdeploy_project.tp.id
+  space_id = octopusdeploy_project.tp.space_id
+  # See https://octopus.com/docs/releases/release-versioning for variable syntax
+  template = "#{Octopus.Version.LastMajor}.#{Octopus.Version.NextMinor}-alpha"
+  depends_on = [
+    octopusdeploy_project_group.tp,
+    octopusdeploy_deployment_process.process
+  ]
+}
+
+resource "octopusdeploy_project_versioning_strategy" "using_donor_package" {
   project_id = octopusdeploy_project.tp.id
   space_id = octopusdeploy_project.tp.space_id
   donor_package_step_id = octopusdeploy_deployment_process.process.step[0].run_script_action[0].id
@@ -80,19 +94,19 @@ resource "octopusdeploy_project_versioning_strategy" "tp" {
 
 ### Required
 
-- `donor_package` (Attributes) Donor Packages. (see [below for nested schema](#nestedatt--donor_package))
 - `project_id` (String) The associated project ID.
-- `space_id` (String) Space ID of the associated project.
 
 ### Optional
 
+- `donor_package` (Attributes) Donor Packages. (see [below for nested schema](#nestedatt--donor_package))
 - `donor_package_step_id` (String) The associated donor package step ID.
+- `space_id` (String) Space ID of the associated project.
 - `template` (String)
 
 <a id="nestedatt--donor_package"></a>
 ### Nested Schema for `donor_package`
 
-Optional:
+Required:
 
 - `deployment_action` (String) Deployment action.
 - `package_reference` (String) Package reference.

--- a/examples/resources/octopusdeploy_project_versioning_strategy/resource.tf
+++ b/examples/resources/octopusdeploy_project_versioning_strategy/resource.tf
@@ -45,7 +45,21 @@ resource "octopusdeploy_deployment_process" "process" {
   depends_on  = [octopusdeploy_project.tp]
 }
 
-resource "octopusdeploy_project_versioning_strategy" "tp" {
+######
+# NOTE: You can use either template or donor_package, not both
+######
+resource "octopusdeploy_project_versioning_strategy" "using_template" {
+  project_id = octopusdeploy_project.tp.id
+  space_id = octopusdeploy_project.tp.space_id
+  # See https://octopus.com/docs/releases/release-versioning for variable syntax
+  template = "#{Octopus.Version.LastMajor}.#{Octopus.Version.NextMinor}-alpha"
+  depends_on = [
+    octopusdeploy_project_group.tp,
+    octopusdeploy_deployment_process.process
+  ]
+}
+
+resource "octopusdeploy_project_versioning_strategy" "using_donor_package" {
   project_id = octopusdeploy_project.tp.id
   space_id = octopusdeploy_project.tp.space_id
   donor_package_step_id = octopusdeploy_deployment_process.process.step[0].run_script_action[0].id

--- a/octopusdeploy_framework/resource_project_versioning_strategy.go
+++ b/octopusdeploy_framework/resource_project_versioning_strategy.go
@@ -2,6 +2,9 @@ package octopusdeploy_framework
 
 import (
 	"context"
+	"log"
+	"net/http"
+
 	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/core"
 	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/packages"
 	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/projects"
@@ -9,8 +12,6 @@ import (
 	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
 	"github.com/hashicorp/terraform-plugin-framework/types"
-	"log"
-	"net/http"
 )
 
 var _ resource.Resource = &projectVersioningStrategyResource{}
@@ -184,7 +185,11 @@ func mapProjectVersioningStrategyToState(versioningStrategy *projects.Versioning
 	if versioningStrategy.DonorPackageStepID != nil {
 		state.DonorPackageStepID = types.StringValue(*versioningStrategy.DonorPackageStepID)
 	}
+	// Template and Donor Package are mutually exclusive options. We won't always have DonorPackage information.
 	state.Template = types.StringValue(versioningStrategy.Template)
-	state.DonorPackage.PackageReference = types.StringValue(versioningStrategy.DonorPackage.PackageReference)
-	state.DonorPackage.DeploymentAction = types.StringValue(versioningStrategy.DonorPackage.DeploymentAction)
+
+	if !(versioningStrategy.DonorPackage == nil) {
+		state.DonorPackage.PackageReference = types.StringValue(versioningStrategy.DonorPackage.PackageReference)
+		state.DonorPackage.DeploymentAction = types.StringValue(versioningStrategy.DonorPackage.DeploymentAction)
+	}
 }

--- a/octopusdeploy_framework/resource_project_versioning_strategy.go
+++ b/octopusdeploy_framework/resource_project_versioning_strategy.go
@@ -14,7 +14,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-var _ resource.Resource = &projectVersioningStrategyResource{}
+var _ resource.ResourceWithConfigValidators = &projectVersioningStrategyResource{}
 
 type projectVersioningStrategyResource struct {
 	*Config
@@ -30,6 +30,10 @@ func (r *projectVersioningStrategyResource) Metadata(_ context.Context, req reso
 
 func (r *projectVersioningStrategyResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
 	resp.Schema = schemas.ProjectVersioningStrategySchema{}.GetResourceSchema()
+}
+
+func (r *projectVersioningStrategyResource) ConfigValidators(context.Context) []resource.ConfigValidator {
+	return schemas.ProjectVersioningStrategySchema{}.GetResourceConfigValidators()
 }
 
 func (r *projectVersioningStrategyResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
@@ -165,20 +169,21 @@ func (r *projectVersioningStrategyResource) Delete(ctx context.Context, req reso
 }
 
 func mapStateToProjectVersioningStrategy(state *schemas.ProjectVersioningStrategyModel) *projects.VersioningStrategy {
-	var donorPackageStepID *string
-	donorPackageStepIDString := state.DonorPackageStepID.ValueString()
-	if donorPackageStepIDString != "" {
-		donorPackageStepID = &donorPackageStepIDString
+	projectVersioningStrategy := &projects.VersioningStrategy{}
+	if !(state.Template.IsNull()) {
+		projectVersioningStrategy.Template = state.Template.ValueString()
 	}
-
-	return &projects.VersioningStrategy{
-		Template:           state.Template.ValueString(),
-		DonorPackageStepID: donorPackageStepID,
-		DonorPackage: &packages.DeploymentActionPackage{
+	if !(state.DonorPackageStepID.IsNull()) {
+		projectVersioningStrategy.DonorPackageStepID = state.DonorPackageStepID.ValueStringPointer()
+	}
+	if !(state.DonorPackage == nil) {
+		projectVersioningStrategy.DonorPackage = &packages.DeploymentActionPackage{
 			DeploymentAction: state.DonorPackage.DeploymentAction.ValueString(),
 			PackageReference: state.DonorPackage.PackageReference.ValueString(),
-		},
+		}
 	}
+
+	return projectVersioningStrategy
 }
 
 func mapProjectVersioningStrategyToState(versioningStrategy *projects.VersioningStrategy, state *schemas.ProjectVersioningStrategyModel) {

--- a/octopusdeploy_framework/schemas/entity_schema.go
+++ b/octopusdeploy_framework/schemas/entity_schema.go
@@ -2,10 +2,16 @@ package schemas
 
 import (
 	datasourceSchema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
 	resourceSchema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
 )
 
 type EntitySchema interface {
 	GetResourceSchema() resourceSchema.Schema
 	GetDatasourceSchema() datasourceSchema.Schema
+}
+
+type EntitySchemaWithResourceValidators interface {
+	EntitySchema
+	GetResourceConfigValidators() []resource.ConfigValidator
 }

--- a/octopusdeploy_framework/schemas/project_versioning_strategy.go
+++ b/octopusdeploy_framework/schemas/project_versioning_strategy.go
@@ -2,7 +2,10 @@ package schemas
 
 import (
 	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator"
 	datasourceSchema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
 	resourceSchema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
 	"github.com/hashicorp/terraform-plugin-framework/types"
@@ -10,7 +13,7 @@ import (
 
 type ProjectVersioningStrategySchema struct{}
 
-var _ EntitySchema = ProjectVersioningStrategySchema{}
+var _ EntitySchemaWithResourceValidators = ProjectVersioningStrategySchema{}
 
 const ProjectVersioningStrategyResourceName = "project_versioning_strategy"
 
@@ -24,7 +27,7 @@ func (p ProjectVersioningStrategySchema) GetResourceSchema() resourceSchema.Sche
 				Build(),
 			"space_id": util.ResourceString().
 				Description("Space ID of the associated project.").
-				Required().
+				Optional().
 				Build(),
 			"donor_package_step_id": util.ResourceString().
 				Description("The associated donor package step ID.").
@@ -35,20 +38,37 @@ func (p ProjectVersioningStrategySchema) GetResourceSchema() resourceSchema.Sche
 				Computed().
 				Build(),
 			"donor_package": resourceSchema.SingleNestedAttribute{
-				Required:    true,
+				Optional:    true,
 				Description: "Donor Packages.",
 				Attributes: map[string]resourceSchema.Attribute{
 					"deployment_action": util.ResourceString().
 						Description("Deployment action.").
-						Optional().
+						Required().
 						Build(),
 					"package_reference": util.ResourceString().
 						Description("Package reference.").
-						Optional().
+						Required().
 						Build(),
 				},
 			},
 		},
+	}
+}
+
+func (p ProjectVersioningStrategySchema) GetResourceConfigValidators() []resource.ConfigValidator {
+	return []resource.ConfigValidator{
+		resourcevalidator.RequiredTogether(
+			path.MatchRoot("donor_package"),
+			path.MatchRoot("donor_package_step_id"),
+		),
+		resourcevalidator.Conflicting(
+			path.MatchRoot("template"),
+			path.MatchRoot("donor_package"),
+		),
+		resourcevalidator.Conflicting(
+			path.MatchRoot("template"),
+			path.MatchRoot("donor_package_step_id"),
+		),
 	}
 }
 
@@ -58,11 +78,11 @@ func (p ProjectVersioningStrategySchema) GetDatasourceSchema() datasourceSchema.
 }
 
 type ProjectVersioningStrategyModel struct {
-	ProjectID          types.String      `tfsdk:"project_id"`
-	SpaceID            types.String      `tfsdk:"space_id"`
-	DonorPackageStepID types.String      `tfsdk:"donor_package_step_id"`
-	Template           types.String      `tfsdk:"template"`
-	DonorPackage       DonorPackageModel `tfsdk:"donor_package"`
+	ProjectID          types.String       `tfsdk:"project_id"`
+	SpaceID            types.String       `tfsdk:"space_id"`
+	DonorPackageStepID types.String       `tfsdk:"donor_package_step_id"`
+	Template           types.String       `tfsdk:"template"`
+	DonorPackage       *DonorPackageModel `tfsdk:"donor_package"`
 }
 
 type DonorPackageModel struct {

